### PR TITLE
[state-sync] Adding timeout logic to peer scoring

### PR DESCRIPTION
**Pull request by sunmilee**
## Motivation

When we choose a peer to receive a chunk for synchronization, we use a scoring system that rewards good actors and punishes bad actors. Peer scoring currently only checks for the peer sending invalid chunk (#837), and this diff adds timeout logic. If a peer fails to respond and is causing a delay in the node making progress, we reduce the peer's score. However, the timed-out peer is not penalized as heavily as sending an invalid chunk error.

**Data structure**: `requests: HashMap<u64, (PeerId, SystemTime)>` in `PeerManager` struct
**Logic**:
1. Whenever a new request for a block is made, its <version, (peer_id, request_time)> is added to a requests hashmap
2. Whenever a response is received, the response for the version is removed from the hashmap
3. During `check_progress` in coordinator.rs, we check if coodinator wasn't able to make progress within timeout limit. We check the current version we are trying to sync, and penalize the peer that is supposed to serve that request. We only penalize if the `current_time - request_time > timeout`
4. We remove the peer from requests map when we penalize the peer to not double penalize later.

## Test Plan

unit test

### CODE REVIEW

File: state_synchronizer/src/coordinator.rs [hunk 1]
```diff
@@ -335,6 +335,9 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
             .map(|x| x.get_value())
             .into_option()
         {
+            // node has received a response from peer, so remove peer entry from requests map
+            self.peer_manager.remove_timeout_request(&version);
+
             if version != self.known_version + 1 {
```
[the remark this answers is not in the document]
Comment (sunmilee) [reply state_synchronizer/src/coordinator.rs]: I'll follow up on next diff!

File: state_synchronizer/src/coordinator.rs [hunk 2]
```diff
@@ -335,6 +335,9 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
             .map(|x| x.get_value())
             .into_option()
         {
+            // node has received a response from peer, so remove peer entry from requests map
+            self.peer_manager.update_timeout_request(version, None);
```
Comment (phoenix-antigravity) [inline state_synchronizer/src/coordinator.rs]: why None? You know peer at this point

File: state_synchronizer/src/peer_manager.rs [hunk 3]
```diff
@@ -135,11 +149,33 @@ impl PeerManager {
...
             .filter(|&(_, peer_info)| peer_info.is_alive && peer_info.is_upstream)
-            .map(|(peer_id, peer_info)| (peer_id, peer_info))
             .collect()
     }
 
     pub fn get_network_sender(&self, peer_id: &PeerId) -> Option<StateSynchronizerSender> {
         self.network_senders.get(peer_id).cloned()
     }
+
+    pub fn update_timeout_request(&mut self, version: u64, peer_id: Option<PeerId>) {
```
Comment (phoenix-antigravity) [inline state_synchronizer/src/peer_manager.rs]: I'd prefer to have 2 separate methods. `process_request` and `process_response`. Although they migh have similar functionality now, it'd be handy to eventually differentiate them

File: state_synchronizer/src/peer_manager.rs [hunk 4]
```diff
@@ -135,11 +149,30 @@ impl PeerManager {
...
     pub fn get_network_sender(&self, peer_id: &PeerId) -> Option<StateSynchronizerSender> {
         self.network_senders.get(peer_id).cloned()
     }
+
+    pub fn process_request(&mut self, version: u64, peer_id: PeerId) {
+        self.requests.insert(version, (peer_id, SystemTime::now()));
+    }
+
+    pub fn process_response(&mut self, version: u64) {
+        self.requests.remove(&version);
```
Comment (phoenix-antigravity) [inline state_synchronizer/src/peer_manager.rs]: should we check here that peer_id is the same?

File: state_synchronizer/src/coordinator.rs [hunk 5]
```diff
@@ -414,6 +420,8 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
                 let mut req = GetChunkRequest::new();
                 req.set_known_version(self.known_version + offset);
                 req.set_limit(self.config.chunk_limit);
+                self.peer_manager
+                    .process_request(self.known_version + offset, peer_id);
```
Comment (phoenix-antigravity) [inline state_synchronizer/src/coordinator.rs]: not sure, but should it be `self.known_version + offset + 1` . Can you double check?

File: state_synchronizer/src/peer_manager.rs [hunk 6]
```diff
@@ -135,11 +149,30 @@ impl PeerManager {
...
+    }
+
+    pub fn process_timeout(&mut self, current_requested_version: u64, timeout: u64) {
+        let request = self.requests.get(&current_requested_version).cloned();
+        if let Some((peer_id, request_time)) = request {
+            if let Ok(request_duration) = SystemTime::now().duration_since(request_time) {
+                if request_duration > Duration::from_millis(timeout) {
+                    self.update_score(&peer_id, PeerScoreUpdateType::TimeOut);
+                    self.process_response(current_requested_version);
```
Comment (phoenix-antigravity) [inline state_synchronizer/src/peer_manager.rs]: nit: it's not necessary process_response call. Just remove it from map

File: state_synchronizer/src/peer_manager.rs [hunk 7]
```diff
@@ -135,11 +149,30 @@ impl PeerManager {
...
+    }
+
+    pub fn process_response(&mut self, version: u64) {
+        self.requests.remove(&version);
+    }
+
+    pub fn process_timeout(&mut self, current_requested_version: u64, timeout: u64) {
+        let request = self.requests.get(&current_requested_version).cloned();
+        if let Some((peer_id, request_time)) = request {
+            if let Ok(request_duration) = SystemTime::now().duration_since(request_time) {
```
Comment (phoenix-antigravity) [inline state_synchronizer/src/peer_manager.rs]: nit: 
  let timeout =  request_time.checked_add(Duration::from_millis(timeout));
  if SystemTime::now().duration_since(tst).is_ok() {...

File: state_synchronizer/src/coordinator.rs (hunk 5 above)
Comment (sunmilee) [reply state_synchronizer/src/coordinator.rs]: Yeah you're right, need the +1, didn't realize offset is always 0. Why is offset always zero?

**Pull request merged**